### PR TITLE
Update GH Action tests and CodeCov

### PR DESCRIPTION
**Pull request by rickecon**
This PR:
- Adds a list of file change event triggers to `build_and_test.yml` so that those tests only run when one of those files is changed.
- Updates the codecov GH Action to version 4 and adds a secret token.
- Adds a list of file change event triggers to `deploy_docs.yml` and `docs_check.yml`, and limits `docs_check.yml` to only run on pull requests.
- Fixes a small typo in `tax_functions.md` in order to test if the event triggers worked properly (yes, they worked)
- Updated some dependencies in `environment.yml`.
- Updated three data files in the `/tests/test_io_data/` file that used output from the taxcalc package. This package was recently updated. I also changed the `test_get_data()` test in the `test_get_micro_data.py` file because the new taxcalc data included four years instead of two years. In order to conserve repo memory footprint, we deleted the last two years of the output.

cc: @jdebacker

### CODE REVIEW

**Comment by rickecon** [comment]
@jdebacker. These three test failures in the last three tests of the `test_get_micro_data.py` Linux Python 3.9 tests seem like legitimate failures.
```
tests/test_get_micro_data.py ........FFF
```

**Comment by rickecon** [comment]
I tried adding the `windows-latest` OS tests back into `build_and_test.yml`, but I got the same error as noted in Issue #78. I am removing the Windows OS tests again from `build_and_test.yml`.
```
Run pytest -m 'not local and not regression' --cov=./ --cov-report=xml
  pytest -m 'not local and not regression' --cov=./ --cov-report=xml
  shell: C:\Program Files\Git\bin\bash.EXE -l {0}
  env:
    INPUT_RUN_POST: true
    CONDA: C:\Users\runneradmin\miniconda[3](https://github.com/PSLmodels/OG-USA/actions/runs/8643695604/job/23698134190?pr=99#step:5:3)
    CONDA_PKGS_DIR: C:\Users\runneradmin\conda_pkgs_dir
C:\Users\runneradmin\miniconda3\envs\ogusa-dev\lib\site-packages\pep8.py:110: FutureWarning: Possible nested set at position 1
  EXTRANEOUS_WHITESPACE_REGEX = re.compile(r'[[({] | []}),;:]')
```
WindowsError

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -1,7 +1,5 @@
-name: Build and test [Python 3.9, 3.10]
-
+name: Build and test
 on: [push, pull_request]
```
Comment (jdebacker) [inline .github/workflows/build_and_test.yml:2]: Re my comments a few weeks ago about unnecessary compute time, from a read of the [updated workflow docs](https://docs.github.com/en/actions), I think we should be able to condition tests to run only when relevant files are affected.

E.g., something like:
```
on:
  [push, pull_request]:
    paths:
      - './ogusa/**.py'
      -  './tests/**.py'
```

Comment (rickecon) [reply .github/workflows/build_and_test.yml:2]: @jdebacker. Yes. I'll add this. But we probably need to expand beyond the `.py` files.

- In the `./ogusa/` directory, we have all `.py` files except for one `.json` file. I think we need to add the `.json` file to the directory tracking.

- In the `./tests/` directory, we probably also need to test changes in the `.pkl` files and `.csv` files in the `./tests/test_io_data/` directory.

- In the main directory, we probably need to run the tests whenever `environment.yml`, `pyproject.toml`, `pytest.ini`, or `setup.py` changes.

File: .github/workflows/docs_check.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name: Check that docs build
-on: [push, pull_request]
+on: [pull_request]
```
[the remark this answers is not in the document]
Comment (jdebacker) [reply .github/workflows/docs_check.yml:2]: Rick cleared this up for me. This should run tests on commits to the PR, but then the deploy will run on the merge (via the `push` action).

**Comment by jdebacker** [comment]
LGTM, thanks @rickecon!  Ready for this to be merged?

**Comment by rickecon** [comment]
@jdebacker. Let me do one last check of a couple things. I still have this white whale of trying to get the Windows tests working again.

**Comment by rickecon** [comment]
@jdebacker. This PR is ready for your final review. I have made all the changes I am going to make. I spent hours trying to get the `windows-latest` tests added back in. But I got everything else in. All these tests should pass by 5pm ET.

**Comment by rickecon** [comment]
@jdebacker. All tests passed locally on my machine.
```
(ogusa-dev) richardevans@Richards-MacBook-Pro-2 OG-USA % pytest
======================= test session starts =========================
platform darwin -- Python 3.11.8, pytest-8.1.1, pluggy-1.4.0
rootdir: /Users/richardevans/Docs/Economics/OSE/OG-USA
configfile: pytest.ini
testpaths: ./tests
plugins: cov-5.0.0, xdist-3.5.0
collected 37 items                                                                                                        

tests/test_calibrate.py ....                                  [ 10%]
tests/test_get_micro_data.py ..............                   [ 48%]
tests/test_income.py ............                             [ 81%]
tests/test_psid_data_setup.py ...                             [ 89%]
tests/test_run_example.py .                                   [ 91%]
tests/test_utils.py .                                         [ 94%]
tests/test_wealth.py ..                                       [100%]
========== 37 passed, 111 warnings in 767.83s (0:12:47) ==================
```

**Pull request merged**